### PR TITLE
First drop

### PR DESCRIPTION
**Pull request by jordigilh**

### CODE REVIEW

File: infra.lvm_tools/shrink_lv/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# ansible-bash-shrink-lv
+Ansible role that runs shrinks a logical volume script. The role contains the shell scripts to shrink the logical volume, as well as the script wrapping it to run as part of the pre-mount step during the boot process.
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/README.md]: Something in the first sentence doesn't work. I'm not sure what you wanted to say here.

File: infra.lvm_tools/shrink_lv/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+## Example of a playbook to run the role
+The following yaml is an example of a playbook that runs the role against a group of hosts named `rhel` and extending its boot partition by 1G. The boot partition is automatically retrieved by the role by identifying the existing mounted partition to `/boot` and passing the information to the script using the `kernel_opts`.
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/README.md]: Maybe rephrase "and extending its" -> "to extend their"

File: infra.lvm_tools/shrink_lv/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      ansible.builtin.setup:
+        gather_subset:
+          - kernel
+  roles:
+    - shrink_lv
+
+```
+
+With an inventory file consisting of group [rhel]:
+```yaml
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/README.md]: This is not a `yaml`, but an `ini` format

File: infra.lvm_tools/shrink_lv/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+A successful execution will look similar to this:
+```bash
+[root@localhost ~]# cat /var/log/messages |grep Resizing -A 2 -B 2
+Oct 16 17:55:00 localhost /dev/mapper/rhel-root: 29715/2686976 files (0.2% non-contiguous), 534773/10743808 blocks
+Oct 16 17:55:00 localhost dracut-pre-mount: resize2fs 1.42.9 (28-Dec-2013)
+Oct 16 17:55:00 localhost journal: Resizing the filesystem on /dev/mapper/rhel-root to 9699328 (4k) blocks.#012The filesystem on /dev/mapper/rhel-root is now 9699328 blocks long.
+Oct 16 17:55:00 localhost journal:  Size of logical volume rhel/root changed from 40.98 GiB (10492 extents) to 37.00 GiB (9472 extents).
+Oct 16 17:55:00 localhost journal:  Logical volume rhel/root successfully resized.
+```
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/README.md]: Add an empty line

File: infra.lvm_tools/shrink_lv/README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
+# ansible-bash-shrink-lv
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/README.md:1]: General comment on this file: Try to break the long lines into multiple lines. Markdown will keep the lines together in the same paragraph as long as there is no empty lines

File: infra.lvm_tools/shrink_lv/galaxy.yml [hunk 6]
```diff
@@ -0,0 +1,68 @@
+### REQUIRED
+# The namespace of the collection. This can be a company/brand/organization or product namespace under which all
+# content lives. May only contain alphanumeric lowercase characters and underscores. Namespaces cannot start with
+# underscores or numbers and cannot contain consecutive underscores
+namespace: infra.lvm_tools
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/galaxy.yml:5]: @swapdisk is this the correct namespace?

File: infra.lvm_tools/shrink_lv/galaxy.yml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+description: Ansible collection to shrink an LV during the pre-mount stage of a booting Linux OS.
+
+# Either a single license or a list of licenses for content inside of a collection. Ansible Galaxy currently only
+# accepts L(SPDX,https://spdx.org/licenses/) licenses. This key is mutually exclusive with 'license_file'
+license:
+- GPL-2.0-or-later
+
+# The path to the license file for the collection. This path is relative to the root of the collection. This key is
+# mutually exclusive with 'license'
+license_file: 'LICENSE'
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/galaxy.yml:27]: As the comment says, you should either set `license` or `license_file` but not both

File: infra.lvm_tools/shrink_lv/meta/runtime.yml [hunk 8]
```diff
@@ -0,0 +1,52 @@
+---
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/meta/runtime.yml]: Since this file is all commented out, you can delete it altogether

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/files/module-setup.sh [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+check(){
+    return 0
+}
+
+install() {
+    inst_multiple -o /usr/bin/numfmt /usr/bin/findmnt /usr/bin/lsblk /usr/sbin/lvm /usr/bin/awk /usr/bin/sed /usr/bin/mktemp /usr/bin/date /usr/bin/head /usr/sbin/blockdev /usr/sbin/tune2fs /usr/sbin/resize2fs /usr/bin/cut /usr/sbin/fsadm
+    inst_hook pre-mount 99 "$moddir/shrink-start.sh"
+    inst_simple "$moddir/fstab" "/etc/fstab.root"
+    inst_simple "$moddir/shrink.sh" "/usr/bin/shrink.sh"
+}
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/roles/shrink_lv/files/module-setup.sh]: Please add a newline at the end of the file

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink-start.sh [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    mv "$tmpfile" /etc/lvm/lvm.conf
+}
+
+
+main() {
+    disable_lvm_lock
+    /usr/bin/shrink.sh --all --fstab=/etc/fstab.root 1>&2 >/dev/kmsg
+}
+
+main "$0"
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink-start.sh]: Please add an empty line

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink.sh [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+        status=$?
+        if [[ $status -ne 0 ]]; then
+            run_status=$status
+        fi
+    done
+
+    exit $run_status
+}
+
+main "$@"
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink.sh]: Please add a newline

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/main.yaml [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+      src: /etc/fstab
+      dest: /usr/lib/dracut/modules.d/99shrink_lv
+      mode: "0444"
+   - name: Create a new initramfs with the extend boot module
+     ansible.builtin.command: /usr/sbin/dracut -a shrink_lv --kver {{ kernel_version }} --force
+     changed_when: true
+   - name: Reboot the server
+     ansible.builtin.reboot:
+      post_reboot_delay: 30
+   - name: Wait for the host to reboot
```
Comment (ygalblum) [inline infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/main.yaml]: Out of curiosity, why do you need this? Doesn't the reboot command cover it? AFAIK before completing, it connect to the remote machine

**Comment by ygalblum** [comment]
In addition, please run [ansible-lint](https://ansible.readthedocs.io/projects/lint/)

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/main.yaml (hunk 12 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/main.yaml]: You're correct, it executes a `whoami` by default, so it should be sufficient.
https://docs.ansible.com/ansible/latest/collections/ansible/builtin/reboot_module.html#parameter-test_command

File: infra.lvm_tools/shrink_lv/meta/runtime.yml (hunk 8 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/meta/runtime.yml]: ACK. I was not sure if there was anything I'd have to keep.

File: infra.lvm_tools/shrink_lv/README.md (hunk 5 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/README.md:1]: ACK.

File: infra.lvm_tools/shrink_lv/galaxy.yml (hunk 7 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/galaxy.yml:27]: ACK

File: infra.lvm_tools/shrink_lv/galaxy.yml (hunk 6 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/galaxy.yml:5]: it seems like `infra.lvm_tools` is the correct namespace, or so @swapdisk told me.

File: infra.lvm_tools/shrink_lv/README.md (hunk 4 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/README.md]: ACK

File: infra.lvm_tools/shrink_lv/README.md (hunk 3 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/README.md]: ACK

File: infra.lvm_tools/shrink_lv/README.md (hunk 2 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/README.md]: Sounds good. Thanks!

File: infra.lvm_tools/shrink_lv/README.md (hunk 1 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/README.md]: Yep, need to reword it.

File: infra.lvm_tools/shrink_lv/README.md (hunk 5 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/README.md:1]: Done.

File: infra.lvm_tools/shrink_lv/README.md (hunk 1 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/README.md]: Done and added additional content related to the option in `/etc/fstab`

File: infra.lvm_tools/shrink_lv/README.md (hunk 2 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/README.md]: Done

File: infra.lvm_tools/shrink_lv/README.md (hunk 3 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/README.md]: Done

File: infra.lvm_tools/shrink_lv/README.md (hunk 4 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/README.md]: Done

File: infra.lvm_tools/shrink_lv/galaxy.yml (hunk 6 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/galaxy.yml:5]: Looks like this name is not compliant with the namespace pattern. This is taken from the [galaxy.json](https://github.com/ansible/ansible-lint/blob/031038ef5e1144d009d55b78a4d1db122ea5ab64/src/ansiblelint/schemas/galaxy.json#L651-L656) validation schema:
```json
 "namespace": {
      "minLength": 2,
      "pattern": "^[a-z][a-z0-9_]+$",
      "title": "Namespace",
      "type": "string"
    },
```

@swapdisk what is your suggestion?

File: infra.lvm_tools/shrink_lv/galaxy.yml (hunk 7 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/galaxy.yml:27]: Removed `license:`

File: infra.lvm_tools/shrink_lv/meta/runtime.yml (hunk 8 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/meta/runtime.yml]: Done

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/files/module-setup.sh (hunk 9 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/roles/shrink_lv/files/module-setup.sh]: Done

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink-start.sh (hunk 10 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink-start.sh]: Done

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink.sh (hunk 11 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink.sh]: Done

**Comment by jordigilh** [comment]
> In addition, please run [ansible-lint](https://ansible.readthedocs.io/projects/lint/)

Done as well

File: infra.lvm_tools/shrink_lv/galaxy.yml (hunk 6 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/galaxy.yml:5]: After discussing with @swapdisk we'll use `infra` as the namespace and will combine both shrink_lv and bigboot roles into the same collection.
@ygalblum once the 2 PRs are merged, I'll open a new PR against your ansible collection for lvm_snapshot to consolidate them.

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/files/module-setup.sh [hunk 13]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+# -*- mode: shell-script; indent-tabs-mode: nil; sh-basic-offset: 4; -*-
+# ex: ts=8 sw=4 sts=4 et filetype=sh
+
+check(){
+    return 0
+}
+
+install() {
+    inst_multiple -o /usr/bin/numfmt /usr/bin/findmnt /usr/bin/lsblk /usr/sbin/lvm /usr/bin/awk /usr/bin/sed /usr/bin/mktemp /usr/bin/date /usr/bin/head /usr/sbin/blockdev /usr/sbin/tune2fs /usr/sbin/resize2fs /usr/bin/cut /usr/sbin/fsadm
```
Comment (swapdisk) [inline infra.lvm_tools/shrink_lv/roles/shrink_lv/files/module-setup.sh]: To make this work with my RHEL7 instances, I needed to also list fsck.ext4 here... 
```suggestion
    inst_multiple -o /usr/bin/numfmt /usr/bin/findmnt /usr/bin/lsblk /usr/sbin/lvm /usr/bin/awk /usr/bin/sed /usr/bin/mktemp /usr/bin/date /usr/bin/head /usr/sbin/blockdev /usr/sbin/tune2fs /usr/sbin/resize2fs /usr/bin/cut /usr/sbin/fsadm /usr/sbin/fsck.ext4
```

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink.sh [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+
+function process_entry() {
+    is_lvm "$1" "$3"
+    status=$?
+    if [[ $status -ne 0 ]]; then
+        return "$status"
+    fi
+    expected_size_in_bytes=$(parse_tag "$2")
+    if [[ -z "$expected_size_in_bytes" ]]; then
+        echo "Error: Tag $SHRINK_TAG not found for device '$3' in '$2'" >&2
```
Comment (swapdisk) [inline infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink.sh]: Let's make this warning rather than error... 
```suggestion
        echo "Warning: Tag $SHRINK_TAG not found for device '$3' in '$2'" >&2
```

**Review by swapdisk — CHANGES_REQUESTED** [review]
Couple suggestions. Otherwise, LGTM.

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink.sh (hunk 14 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink.sh]: Done

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/files/module-setup.sh (hunk 13 above)
Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/roles/shrink_lv/files/module-setup.sh]: Done
